File: calendar-server/src/main/java/uk/co/unclealex/hammers/calendar/server/html/SeasonHtmlGamesScanner.java
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 */

package uk.co.unclealex.hammers.calendar.server.html;

import java.io.IOException;
import java.net.URI;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.Arrays;
import java.util.Iterator;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.htmlcleaner.TagNode;
import org.htmlcleaner.XPatherException;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import uk.co.unclealex.hammers.calendar.server.dates.UnparseableDateException;
import uk.co.unclealex.hammers.calendar.server.model.GameKey;
import uk.co.unclealex.hammers.calendar.shared.model.Competition;
import uk.co.unclealex.hammers.calendar.shared.model.Location;

import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.common.base.Strings;
import com.google.common.collect.Iterators;

/**
 * An {@link HtmlGamesScanner} that scans the season's fixtures page for game
 * information.
 * 
 * @author alex
 * 
 */
public class SeasonHtmlGamesScanner extends StatefulDomBasedHtmlGamesScanner {

  /** The logger for this class. */
  private static final Logger log = LoggerFactory.getLogger(SeasonHtmlGamesScanner.class);

  /**
   * {@inheritDoc}
   */
  @Override
  protected Scanner createScanner(final URI uri, final TagNode tagNode) {
    return new SeasonScanner(uri, tagNode);
  }

  /**
   * The {@link Scanner} that scans the season's fixtures page for game
   * information.
   * 
   * @author alex
   * 
   */
  class SeasonScanner extends Scanner {

    /**
     * The current season.
     */
    private int season;

    /**
     * The current month.
     */
    private String month;

    /**
     * The {@link DateTime} the season started.
     */
    private DateTime startOfSeason;

    /**
     * Instantiates a new season scanner.
     * 
     * @param uri
     *          the uri
     * @param tagNode
     *          the tag node
     */
    public SeasonScanner(final URI uri, final TagNode tagNode) {
      super(uri, tagNode);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void scan() throws IOException {
      updateSeason();
      TagNode tableTagNode;
      try {
        tableTagNode = (TagNode) getTagNode().evaluateXPath("//table[@class='fixtureList']")[0];
      }
      catch (final XPatherException e) {
        throw new IOException(e);
      }
      final TagNodeFilter tableRowFilter = new TagNodeFilter() {

        @Override
        public boolean apply(final TagNode tagNode) {
          return "tr".equals(tagNode.getName());
        }
      };
      class ClassContainsPredicate implements Predicate<TagNode> {
        String target;

        public ClassContainsPredicate(final String target) {
          super();
          this.target = target;
        }

        @Override
        public boolean apply(final TagNode tagNode) {
          final String classes = tagNode.getAttributeByName("class");
          return classes != null && classes.contains(target);
        }
      }
      final Predicate<TagNode> isMonthRowPredicate = new ClassContainsPredicate("rowHeader");
      final Predicate<TagNode> isGameRow = new ClassContainsPredicate("fixture");
      for (final TagNode row : tableRowFilter.list(tableTagNode)) {
        if (isMonthRowPredicate.apply(row)) {
          updateMonth(row);
        }
        else if (isGameRow.apply(row)) {
          updateGame(row);
        }
      }
    }

    /**
     * Find which season page represents.
     */
    protected void updateSeason() {
      final Pattern seasonPattern = Pattern.compile("s\\.prop3=\"([0-9]+)\"");
      new TagNodeWalker(getTagNode()) {

        @Override
        public void execute(final TagNode tagNode) {
          if (getSeason() == 0 && "script".equals(tagNode.getName())) {
            final String text = TagNodeUtils.normaliseText(tagNode);
            final Matcher matcher = seasonPattern.matcher(text);
            if (matcher.find()) {
              final int season = Integer.valueOf(matcher.group(1));
              log.info("Found season " + season);
              setSeason(season);
              setStartOfSeason(getDateService().parseDate("01/07/" + season, "dd/MM/yyyy"));
            }
          }
        }
      };
    }

    /**
     * Update the month.
     * 
     * @param row
     *          The table row containing the month.
     */
    protected void updateMonth(final TagNode row) {
      final TagNode child = row.findElementByName("td", false);
      final String month = TagNodeUtils.normaliseText(child);
      log.info("Found " + month + " " + getSeason());
      setMonth(month);
    }

    /**
     * Update a game.
     * 
     * @param row
     *          The current row in the fixtures table.
     */
    protected void updateGame(final TagNode row) {
      final Iterator<TagNode> tds = Iterators.forArray(row.getElementsByName("td", false));
      String date = TagNodeUtils.normaliseTextToNull(tds.next());
      date = date.replaceAll("[^0-9]", "");
      date = Strings.padStart(date, 2, '0');
      final String time = TagNodeUtils.normaliseTextToNull(tds.next());

      final Location location =
          "H".equals(TagNodeUtils.normaliseTextToNull(tds.next())) ? Location.HOME : Location.AWAY;
      TagNode opponentsEl = tds.next();
      // Could be in a link
      final TagNode opponentsLink = opponentsEl.findElementByName("a", false);
      if (opponentsLink != null) {
        opponentsEl = opponentsLink;
      }
      final String opponents = TagNodeUtils.normaliseTextToNull(opponentsEl);
      final Competition competition = Competition.findByToken(TagNodeUtils.normaliseTextToNull(tds.next()));
      final GameKey gameKey = new GameKey(competition, location, opponents, getSeason());
      log.info("Found game key " + gameKey);
      final String datePlayedString = Joiner.on(" ").join(date, time, getMonth());
      DateTime datePlayed;
      try {
        datePlayed =
            getDateService().parsePossiblyYearlessDate(
                datePlayedString,
                getStartOfSeason(),
                false,
                new String[] { ("dd HH:mm MMMM[ yyyy]") });
      }
      catch (final UnparseableDateException e) {
        log.warn("Cannot parse date " + datePlayedString + " for game " + gameKey, e);
        return;
      }
      final GameLocator gameKeyLocator = GameLocator.gameKeyLocator(gameKey);
      final GameUpdateCommand datePlayedGameUpdateCommand = GameUpdateCommand.datePlayed(gameKeyLocator, datePlayed);
      tds.next(); // Move past the W/L/D token.
      final GameUpdateCommand resultGameUpdateCommand =
          GameUpdateCommand.result(gameKeyLocator, TagNodeUtils.normaliseTextToNull(tds.next()));
      final String attendenceText = TagNodeUtils.normaliseTextToNull(tds.next());
      Integer attendence;
      final URI uri = getUri();
      try {
        attendence =
            (attendenceText == null || "00".equals(attendenceText)) ? null : NumberFormat
                .getIntegerInstance(Locale.UK)
                .parse(attendenceText)
                .intValue();
      }
      catch (final ParseException e) {
        final StringBuilder sb = new StringBuilder();
        for (final byte by : attendenceText.getBytes()) {
          int i;
          if (by < 0) {
            i = 256 + by;
          }
          else {
            i = by;
          }
          sb.append("0x").append(Integer.toHexString(i));
        }
        log.warn("Cannot parse attendance " + sb.toString() + " on page " + uri, e);
        attendence = null;
      }
      final GameUpdateCommand attendenceUpdateCommand = GameUpdateCommand.attendence(gameKeyLocator, attendence);
      tds.next(); // Move past the league table token.
      final TagNode matchReportTd = tds.next();
      final TagNode matchReportLink = matchReportTd.findElementByName("a", false);
      String matchReport;
      if (matchReportLink != null) {
        final String matchReportPath = matchReportLink.getAttributeByName("href");
        final URI matchReportUri = uri.resolve(matchReportPath);
        matchReport = matchReportUri.toString();
      }
      else {
        matchReport = null;
      }
      final GameUpdateCommand matchReportUpdateCommand = GameUpdateCommand.matchReport(gameKeyLocator, matchReport);
      getGameUpdateCommands().addAll(
          Arrays.asList(
              datePlayedGameUpdateCommand,
              resultGameUpdateCommand,
              attendenceUpdateCommand,
              matchReportUpdateCommand));
    }

    /**
     * Gets the current season.
     * 
     * @return the current season
     */
    public int getSeason() {
      return season;
    }

    /**
     * Gets the current month.
     * 
     * @return the current month
     */
    public String getMonth() {
      return month;
    }

    /**
     * Sets the current month.
     * 
     * @param month
     *          the new current month
     */
    public void setMonth(final String month) {
      this.month = month;
    }

    /**
     * Gets the {@link DateTime} the season started.
     * 
     * @return the {@link DateTime} the season started
     */
    public DateTime getStartOfSeason() {
      return startOfSeason;
    }

    /**
     * Sets the {@link DateTime} the season started.
     * 
     * @param startOfSeason
     *          the new {@link DateTime} the season started
     */
    public void setStartOfSeason(final DateTime startOfSeason) {
      this.startOfSeason = startOfSeason;
    }

    /**
     * Sets the current season.
     * 
     * @param season
     *          the new current season
     */
    public void setSeason(final int season) {
      this.season = season;
    }
  }
}

File: app/uk/co/unclealex/hammers/calendar/html/SeasonHtmlGamesScanner.scala
/**
 * Copyright 2010-2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package uk.co.unclealex.hammers.calendar.html;

import java.net.URI
import java.text.NumberFormat
import java.text.ParseException
import java.util.Locale
import org.htmlcleaner.TagNode
import org.joda.time.DateTime
import org.slf4j.Logger
import TagNodeImplicits._
import uk.co.unclealex.hammers.calendar.dates.DateService
import uk.co.unclealex.hammers.calendar.dates.UnparseableDateException
import uk.co.unclealex.hammers.calendar.model.Competition
import uk.co.unclealex.hammers.calendar.model.GameKey
import uk.co.unclealex.hammers.calendar.model.Location
import javax.inject.Inject

/**
 * An {@link HtmlGamesScanner} that scans the season's fixtures page for game
 * information.
 *
 * @author alex
 *
 */
class SeasonHtmlGamesScanner @Inject() (
  /**
   * The {@link HtmlPageLoader} used to load web pages.
   */
  htmlPageLoader: HtmlPageLoader,
  /**
   * The {@link DateService} to use for date and time manipulation.
   */
  dateService: DateService) extends StatefulDomBasedHtmlGamesScanner(htmlPageLoader, dateService) {

  protected override def createScanner(uri: URI, tagNode: TagNode) =
    new SeasonScanner(uri, tagNode)

  /**
   * The {@link Scanner} that scans the season's fixtures page for game
   * information.
   *
   * @author alex
   *
   */
  class SeasonScanner(uri: URI, tagNode: TagNode)
    extends Scanner(uri, tagNode) {

    /**
     * The current season.
     */
    var season: Option[Integer] = None

    /**
     * The current month.
     */
    var month: Option[String] = None

    /**
     * The {@link DateTime} the season started.
     */
    var startOfSeason: Option[DateTime] = None

    def scan: List[GameUpdateCommand] = {
      updateSeason
      val tableTagNode = tagNode.evaluateXPath("//table[@class='fixtureList']")(0).asInstanceOf[TagNode]
      val tableRowFilter = new TagNodeFilter(tg => "tr" == tg.getName)

      val classContainsPredicate = { className: String =>
        tagNode: TagNode =>
          val classes = tagNode.getAttributeByName("class")
          classes != null && classes.contains(className)
      }
      val isMonthRowPredicate = classContainsPredicate("rowHeader")
      val isGameRow = classContainsPredicate("fixture")
      tableRowFilter.list(tableTagNode).flatMap { row =>
        if (isMonthRowPredicate(row)) {
          updateMonth(row)
          List.empty
        } else if (isGameRow(row)) {
          updateGame(row)
        } else {
          List.empty
        }
      }
    }

    /**
     * Find which season page represents.
     */
    def updateSeason: Unit = {
      val seasonPattern = """s\.prop3="([0-9]+)""".r.unanchored
      TagNodeWalker.walk { tagNode =>
        if (season.isEmpty && "script" == tagNode.getName) {
          tagNode.normalisedText match {
            case seasonPattern(year) => {
              logger info s"Found season $year"
              season = Some(Integer.parseInt(year))
              startOfSeason = dateService.parseDate(s"01/07/$year", "dd/MM/yyyy")
            }
            case _ => // Ignore the line
          }
        }
        None
      }(tagNode)
    }

    /**
     * Update the month.
     *
     * @param row
     *          The table row containing the month.
     */
    def updateMonth(row: TagNode): Unit = {
      val child = row.findElementByName("td", false);
      val month = child.normalisedText
      logger info s"Found $month ${season.get}"
      this.month = Some(month)
    }

    def padStart(str: String)(implicit minLength: Int, padChar: Char): String = {
      if (str.length < minLength) {
        padStart(padChar + str)
      } else {
        str
      }
    }
    /**
     * Update a game.
     *
     * @param row
     *          The current row in the fixtures table.
     */
    def updateGame(row: TagNode): List[GameUpdateCommand] = {
      val tds: Iterator[TagNode] = row.getElementsByName("td", false).toSeq.iterator
      val date = padStart(tds.next.normalisedText.replaceAll("[^0-9]", ""))(2, '0')
      val time = tds.next.normalisedText
      val location = if ("H" == tds.next.normalisedText) Location.HOME else Location.AWAY
      val opponentsElOrLink = tds.next
      // Could be in a link
      val opponentsEl = Option(opponentsElOrLink.findElementByName("a", false)).getOrElse(opponentsElOrLink)
      val opponents = opponentsEl.normalisedText
      val competition = Competition.findByToken(tds.next.normalisedText)
      val gameKey = new GameKey(competition, location, opponents, season.get)
      logger info s"Found game key $gameKey"
      val datePlayedString = List(date, time, month.get).mkString(" ")
      dateService.parsePossiblyYearlessDate(
        datePlayedString,
        startOfSeason.get,
        false,
        "dd HH:mm MMMM[ yyyy]") match {
          case None => {
            logger warn s"Cannot parse date $datePlayedString for game $gameKey"
            List.empty
          }
          case Some(datePlayed) => {
            val gameKeyLocator = GameKeyLocator(gameKey)
            val datePlayedGameUpdateCommand: Option[GameUpdateCommand] =
              Some(DatePlayedUpdateCommand(gameKeyLocator, datePlayed))
            tds.next(); // Move past the W/L/D token.
            val resultGameUpdateCommand =
              tds.next.optionalNormalisedText map (text => ResultUpdateCommand(gameKeyLocator, text))
            val attendenceText = tds.next().normalisedText
            val attendence: Option[Int] = {
              try {
                if (attendenceText.isEmpty || "00" == attendenceText) None else {
                  Some(NumberFormat.getIntegerInstance(Locale.UK)
                    .parse(attendenceText)
                    .intValue())
                }
              } catch {
                case e: ParseException => {
                  val invalidData =
                    attendenceText.getBytes.map(by => (if (by < 0) 256 + by else by).asInstanceOf[Int]).map(i =>
                      "0x" + Integer.toHexString(i)).mkString
                  logger warn (s"Cannot parse attendance $invalidData on page $uri", e)
                  None
                }
              }
            }
            val attendenceUpdateCommand =
              attendence map (attendence => AttendenceUpdateCommand(gameKeyLocator, attendence))
            tds.next(); // Move past the league table token.
            val matchReportTd = tds.next();
            val matchReportLink = Option(matchReportTd.findElementByName("a", false))
            val matchReport = matchReportLink.map { matchReportLink =>
              val matchReportPath = matchReportLink.getAttributeByName("href")
              val matchReportUri = uri.resolve(matchReportPath)
              matchReportUri.toString
            }
            val matchReportUpdateCommand =
              matchReport map (matchReport => MatchReportUpdateCommand(gameKeyLocator, matchReport))
            val newUpdateCommands =
              List.empty ++ datePlayedGameUpdateCommand ++ resultGameUpdateCommand ++ attendenceUpdateCommand ++
                matchReportUpdateCommand
            logger debug s"Adding game update commands $newUpdateCommands"
            newUpdateCommands
          }
        }
    }
  }
}
